t/SocketWithLogging.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <boost/asio/buffers_iterator.hpp>
+#include <boost/system/error_code.hpp>
+#include <iostream>
+#include <utility>
+
+namespace bredis {
+namespace test {
+
+template <typename NextLayer> class SocketWithLogging {
+    NextLayer stream_;
+
+  public:
+    template <typename... Args>
+    SocketWithLogging(Args &&... args) : stream_(std::forward<Args>(args)...) {}
+
+    void cancel() {
+        std::cout << "{cancel}" << std::endl;
+        stream_.cancel();
+    }
+
+    template <typename BufferSequence>
+    void dump(const char *prefix, const BufferSequence &buffers) {
+        using Iterator = boost::asio::buffers_iterator<BufferSequence, char>;
+
+        Iterator it = Iterator::begin(buffers);
+        Iterator end = Iterator::end(buffers);
+        auto size = std::distance(it, end);
+
+        std::cout << "{" << prefix << " " << size << " bytes}[";
+        for (; it != end; it++) {
+            std::cout << *it;
+        }
+        std::cout << "]" << std::endl;
+    }
+
+    template <typename MutableBufferSequence, typename ReadHandler>
+    inline void async_read_some(const MutableBufferSequence &buffers,
```

Comment:
`inline` is redundant, member function definitions inside class declarations are automatically marked `inline`. You need to think of the `inline` keyword as "multiple definitions allowed" rather than "insert the body of the function into each call site."